### PR TITLE
PERF-#6437: preserve dtypes for 'reindex'

### LINKED ISSUE #6437: Preserve dtypes cache for `reindex`

### PR DESCRIPTION
**Pull request by anmyachev**
## What do these changes do?

- [x] first commit message and PR title follow format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
  > **_NOTE:_**  If you edit the PR title to match this format, you need to add another commit (even if it's empty) or amend your last commit for the CI job that checks the PR title to pick up the new PR title.
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #6437 
- [x] tests added and passing
- [x] module layout described at `docs/development/architecture.rst` is up-to-date

### CODE REVIEW

File: modin/core/storage_formats/pandas/query_compiler.py [hunk 1]
```diff
@@ -640,13 +640,30 @@ def map_func(left, right=right_pandas, kwargs=kwargs):  # pragma: no cover
...
         new_columns, _ = self.columns.reindex(labels) if axis else (self.columns, None)
+        new_dtypes = None
+        if (
+            self._modin_frame.has_materialized_dtypes
+            and kwargs.get("method", None) is None
+        ):
+            dtype = pandas.Index([kwargs.get("fill_value", np.nan)]).dtype
+            if axis == 0:
+                new_dtypes = self.dtypes.copy()
+                if indexer is not None and -1 in indexer:
```
Comment (YarShev) [inline modin/core/storage_formats/pandas/query_compiler.py:653]: What is the magic number equal to -1? Let's assing it to a descriptive var.

File: modin/core/storage_formats/pandas/query_compiler.py [hunk 2]
```diff
@@ -640,13 +640,30 @@ def map_func(left, right=right_pandas, kwargs=kwargs):  # pragma: no cover
...
+                    # case where missing rows are filled with `fill_value` and can change column type
+                    from pandas.core.dtypes.cast import find_common_type
```
Comment (YarShev) [inline modin/core/storage_formats/pandas/query_compiler.py:655]: Move to the top if this method is really needed.

File: modin/core/storage_formats/pandas/query_compiler.py [hunk 3]
```diff
@@ -640,13 +640,30 @@ def map_func(left, right=right_pandas, kwargs=kwargs):  # pragma: no cover
...
     def reindex(self, axis, labels, **kwargs):
-        new_index, _ = (self.index, None) if axis else self.index.reindex(labels)
+        new_index, indexer = (self.index, None) if axis else self.index.reindex(labels)
         new_columns, _ = self.columns.reindex(labels) if axis else (self.columns, None)
+        new_dtypes = None
+        if (
+            self._modin_frame.has_materialized_dtypes
+            and kwargs.get("method", None) is None
+        ):
+            dtype = pandas.Index([kwargs.get("fill_value", np.nan)]).dtype
```
[the remark this answers is not in the document]
Comment (anmyachev) [reply modin/core/storage_formats/pandas/query_compiler.py:650]: Index is a simpler entity than Series, and it suffices here.

File: modin/core/storage_formats/pandas/query_compiler.py [hunk 4]
```diff
@@ -640,13 +640,30 @@ def map_func(left, right=right_pandas, kwargs=kwargs):  # pragma: no cover
...
+            if axis == 0:
+                new_dtypes = self.dtypes.copy()
+                if indexer is not None and -1 in indexer:
+                    # case where missing rows are filled with `fill_value` and can change column type
+                    from pandas.core.dtypes.cast import find_common_type
+
+                    for col in new_dtypes.keys():
+                        new_dtypes[col] = find_common_type((new_dtypes[col], dtype))
```
[the remark this answers is not in the document]
Comment (anmyachev) [reply modin/core/storage_formats/pandas/query_compiler.py:658]: I'm not sure, but a few parts of the code still use the function from pandas (so I decided to use it too).

File: modin/core/storage_formats/pandas/query_compiler.py [hunk 5]
```diff
@@ -640,13 +640,29 @@ def map_func(left, right=right_pandas, kwargs=kwargs):  # pragma: no cover
...
+            and kwargs.get("method", None) is None
+        ):
+            dtype = pandas.Index([kwargs.get("fill_value", np.nan)]).dtype
+            if axis == 0:
+                new_dtypes = self.dtypes.copy()
+                # "-1" means that the required labels are missing in the dataframe and the
+                # corresponding rows will be filled with "fill_value" that may change the column type.
+                if indexer is not None and -1 in indexer:
+                    for col in new_dtypes.keys():
+                        new_dtypes[col] = find_common_type((new_dtypes[col], dtype))
```
Comment (vnlitvinov) [inline modin/core/storage_formats/pandas/query_compiler.py:657]:
```suggestion
                    for col, col_dtype in new_dtypes.items():
                        new_dtypes[col] = find_common_type((col_dtype, dtype))
```

**Review by vnlitvinov — COMMENTED** [review]
Overall looks good, not stamping approval because I still would like to see the in-code comment about columns vs rows added first 😺

File: modin/core/storage_formats/pandas/query_compiler.py [hunk 6]
```diff
@@ -640,13 +640,30 @@ def map_func(left, right=right_pandas, kwargs=kwargs):  # pragma: no cover
 
     # Reindex/reset_index (may shuffle data)
     def reindex(self, axis, labels, **kwargs):
-        new_index, _ = (self.index, None) if axis else self.index.reindex(labels)
+        new_index, indexer = (self.index, None) if axis else self.index.reindex(labels)
         new_columns, _ = self.columns.reindex(labels) if axis else (self.columns, None)
```
[the remark this answers is not in the document]
Comment (anmyachev) [reply modin/core/storage_formats/pandas/query_compiler.py:644]: > Please add this as an in-code comment 🙃

@vnlitvinov added

**Review by vnlitvinov — APPROVED** [review]
LGTM! Let's wait for @YarShev feedback before merging, though, because he also left some questions and other feedback.

**Review by YarShev — APPROVED** [review]

**Pull request merged**